### ISSUE TITLE
Server creation fails (bad server.url) on Travis CI for NodeJS v11 and io.js v1.2.0

### ISSUE THREAD

**Issue by ArminTamzarian** [mcavage/node-restify#751]
Myy project [node-misfit](https://github.com/ArminTamzarian/node-misfit) uses node-restify to build a [mock service](https://github.com/ArminTamzarian/node-misfit/blob/master/test/mock-api.misfitwearables.com.js) for unit testing. While running just fine locally (OSX - NodeJS v0.10.29) when I push the current in progress code to Github Travis CI tests the build with what appears to be an invalid URL detected.

Specifically my code for creating the server is as follows:
```javascript
var mock = function (options, callback) {
          var server = restify.createServer();
          server.use(restify.bodyParser());

          // routing logic...

          server.listen(options.port, function() {
            console.log(server); // for debugging this issue
            console.log(server.url); // for debugging this issue
            callback();
        });
```

While I am testing on NodeJS v0.11.x, v0.10.x and io.js v1.2.x my tests only pass on NodeJS v0.10.x (which makes since as the current version of io.js was forked from the NodeJS v0.11 branch). On startup I am outputting the URL created by the server startup, and the results are as follows:

* **NodeJS v0.11** - server.url -> `http://:::8000`
* **NodeJS v0.10** - server.url -> `http://0.0.0.0:8000`
* **io.js v1.2** - server.url -> `http://:::8000`

For more information see the Travis CI output for the referenced build: https://travis-ci.org/ArminTamzarian/node-misfit/builds/50702849

So, I'm not sure if there is an issue between restify and the current configuration of Travis CI or if there is a bug between restify and the NodeJS v0.11 branch (or if I am doing something egregiously wrong). Any assistance would be appreciated.

**Comment by ArminTamzarian** [comment]
No sooner did I report this than I installed nvm and tested locally with v0.11. Same issue as Travis CI, but I drilled down into the issue and found that the call to `server.listen(port, callback)` tried to bind to an nonexistent IPv6 address `:::` by default whereas on v0.10 the same call binds to the IPv4 address `0.0.0.0` (all addresses).

When I changed this call to the form `server.listen(port, '127.0.0.1', callback)` all versions bound to the IPv4 address of 127.0.0.1 correctly and all tests [pass](https://travis-ci.org/ArminTamzarian/node-misfit/builds/50780077). I'm still not sure if this is an issue caused from NodeJS v0.11 or some interaction between v0.11 and Restify so I will leave this open for someone more knowledgable on the situation to comment/close.

**Comment by ArminTamzarian** [comment]
Okay, taking a moment to check it looks like this is indeed a NodeJS v11 (and v12) issue.

Running a test script of:

```javascript
var server = http.createServer();

server.listen(8000, function() {
    console.log(server.address());
});
```

The results are as follows:

* **v0.10.x** server.address() -> `{ address: '0.0.0.0', family: 'IPv4', port: 8000 }`
* **v0.11.x** server.address() -> `{ address: '::', family: 'IPv6', port: 8000 }`
* **v0.12.x** server.address() -> `{ address: '::', family: 'IPv6', port: 8000 }`

So despite the [docs](http://nodejs.org/api/http.html#http_server_listen_port_hostname_backlog_callback) stating `If the hostname is omitted, the server will accept connections directed to any IPv4 address (INADDR_ANY)` the reality is incorrect.

tl;dr: Not a Restify issue but rather a NodeJS v0.11/0.12 issue. Closing.

**Comment by ArminTamzarian** [comment]
One further note. Apparently `INADDR_ANY` for IPv4 is `0.0.0.0` whereas for IPv6 is is `::`. Therefore the address is right and the whole issue may reside in the implementation of the superagent library instead. Should have figured that the NodeJS engineers wouldn't have ended up with such a glaring issue that would remain unreported/unresolved.

mea culpa

**Comment by ArminTamzarian** [comment]
After finally taking a little bit more time to look into IPv6 it turns out that everything seems to work functionally across all modules except for one small issue with Restify. Specifically, the defined `url` property does not properly take into account the bracketing requirements for IPv6 addresses when appending ports. As a result an HTTP server created on `::1` or `::` should return a url of `http://[::1]:80` and `http://[::]:80` respectively.
